neta.1
.\" @(#)man page for netl network logger/sniffer
.\" copyright 1997 graham the ollis
.ad b
.TH neta 1 "3 june 1997" "CORE software" "CORE software"
.AT 3
.de sh
.br
.ne 5
.PP
\fB\\$1\fR
.PP
..
.SH NAME
neta \- network packet analizer
.SH SYNOPSIS
.B neta
[
.B -v-
] [
.B -r-
] [
.B -h
] file [...]
.SH DESCRIPTION
.I neta
is a network packet analizer.  it was designed to decode an IP packet 
stored as a file.  usually this will be a packet dumped by 
.BR netl (8).
first the address is shown, followed by header data.  the final part is
a hex dump, which displays offset, hex data, and character data.
because the header data is interpreted for you, it is not included in
the hex dump.
.SH OPTIONS
all boolean options can be postpended with a \+ or \- in order turn them
on or off.  the default invokation is always +, but the default value may
vary.
.PP
.TP 8
.TP
.BR \-h
help.  print out usage information.
.PP
.TP
.BR \-r [+|-]
resolve IP numbers to their hostname (default is on).
.PP
.TP
.BR \-v [+|-]
display version number and copyright information (default is on).
.SH SEE ALSO
.BR dcp (1),
.BR xd (1),
.BR netl (8),
.BR hwpassive (8)
.SH BUGS
almost certainly.  please send bug reports/fixes to me 
Graham THE Ollis <ollisg@ns.arizona.edu>
.SH COPYING
Copyright 1996, 1997 Graham THE Ollis
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.SH AUTHOR
Graham THE Ollis <ollisg@ns.arizona.edu>


HISTORY
version .95.2
-------------

 . mostly internal stuff.  added parse.c and grab.c modules, which contain
   code that used to be in the netl.c main module.  this is so that all the 
   network specific code is in grab.c, the true nightmare for porting is
   in this file... so if i can port grab.c to win32... then i'm done.

 . the \ character now works in the config file, if you want to have
   multiple line rules.  very unix-esq.

version .95
-----------

 . created man page documentation.

 . added `detect' config line, see the netl(8) manpage for details.
   this will reduce the amount of configeration you will have to do
   in the netl.conf file.

 . modified dist version of netl.conf so that it's more portable.
   in particular, i put in literal port numbers rather than using
   port names.

 . dcp.pl uses perl native Socket module instead of using net cat.
   `dcp.pl -o' uses the old method.

version .94
-----------

 . changed all the struct declerations to typedef declerations.  get off my
   back ok?  i've been programming perl for so long, forgot some of that
   strange C language. 

 . no longer depending on linux .h files for IP (and sub) headers.  this
   is for portability reasons.  if you are doing a port, make sure you get
   the endian type right!  there are now three global headers:
	global.h - general global stuff (endian integer conversion etc.)
	ether.h - ethernet specific data types, #defines
	ip.h - ip protocol specific data types, #defines
   logically, if your adding an interface (ATM, token ring, ppp, etc.)
   you'd want to start in ether.h.  if your adding an OS, look at global.h
   first.  if you want to add a protocol (IPX, etc.) start with ip.h and
   modify/add to these files.  if your trying to do all of this at once,
   god help you. 

 . fixed the clearipcache() function.  thus now, when you send a 
   netl:readconfig, netl will clear the ip cache before reading the file,
   avoiding all sorts of potential duplicates.  though you probably have
   an enourmous amount of memory like me and don't care... 

 . added some global options with ports to win32 in mind.  NO_SYSLOGD, if 
   defined in global.h will not even compile with the syslogd library stuff.  
   this is a good idea if your running from a non unix system.  note that 
   syslogd is really nice in that it partially compresses lines by indicating 
   repeated lines.  this won't save you if you get strobed, but will keep your 
   log file neat if some one ping -f's you from only one site.

 . added NO_TEEOUT option which is defined for unix.  the -o option should 
   only be necessary in win32 only.  even then, you should only compile with
   this option if your thinking of using it.  -o copies output to the specified
   file.

0.93 (and 0.92)
---------------

 a. -z option allows you to run netl in the foreground.  output is sent to
    stdout instead of syslogd.

 b. you can include configeration lines in the command line.  when you do
    this, the config file /etc/netl.conf will NOT be read.  note that (a)
    and (b) are meant to be used in concert with each other for testing
    and quick monitoring. 

 c. dcp is partially implemented.  netl will listen to port 47 udp
    (silently) if you `listen 47' in your config file.  when you send the
    message `netl:readconfig' netl is suppose to re read your config file. 
    dcp.pl is a perl script (you'll need net cat for this to work) which
    sends this message.  what is left to do for dcp?
	1. generic server side dcp interface libdcp.a file or somesuch.
	2. more solid C version of dcp client.

 d. netl is now officially under the GPL, see COPYING for more information.

 e. icmp codes are now included in some .h file somewhere rather than
    depending on your linux being overly up to date.

 f. neta no longer swaps source/destination ip/ethernet addresses
    inapropriately. 

 g. started work on hwpassive, which is no where near complete.

 h. tdr.pl is no included with netl distribution for convenience.  get
    your own version at: 

http://www.u.arizona.edu/~ollisg/core.html
	[ CORE ]

netl 0.91
---------

 minor updates

netl 0.90
---------

 initial release


neta.c
/*==============================================================================
| neta
|   code by Graham THE Ollis <ollisg@netl.org>
|
|   Copyright (C) 1997 Graham THE Ollis <ollisg@netl.org>
|
|   This program is free software; you can redistribute it and/or modify
|   it under the terms of the GNU General Public License as published by
|   the Free Software Foundation; either version 2 of the License, or
|   (at your option) any later version.
|
|   This program is distributed in the hope that it will be useful,
|   but WITHOUT ANY WARRANTY; without even the implied warranty of
|   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
|   GNU General Public License for more details.
|
|   You should have received a copy of the GNU General Public License
|   along with this program; if not, write to the Free Software
|   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
|
|  Date       Name	Revision
|  ---------  --------  --------
|  23 Feb 97  G. Ollis	.92 created network analysis software
|  28 Feb 97  G. Ollis	started using the net32 conversion for 32 bit integers
|			moved read() and dumpdata() to dump.c
|=============================================================================*/

static char	*id = "@(#)neta (c) 1997 graham the ollis <ollisg@netl.org>";
static void dumb(char *c) { dumb(id); }

#include <stdio.h>
#include <stdlib.h>

#include "netl/global.h"
#include "netl/ether.h"
#include "netl/ip.h"

#include "netl/dump.h"
#include "netl/lookup.h"
#include "netl/options.h"
#include "netl/resolve.h"

/*==============================================================================
| prototypes
|=============================================================================*/

void print(u8 *, size_t);
size_t printip(u8 *);
void printaddrport(u8 *, size_t);
void printtcp(u8 *, size_t);
void printicmp(u8 *, size_t);
void printaddr(iphdr *);

/*==============================================================================
| main
| + this is the neta(1) main routine.  see the man page for details on its use.
|=============================================================================*/

int
main(int argc, char *argv[])
{
	int		i;
	u8		*buff;
	size_t	size;

	prog = argv[0];

	parsecmdline(argc, argv);
	if(displayVersion) {
		fputs("neta ", stdout);
		puts(COPYVER);
	}

	if(argc < 2) {
		fprintf(stderr, "usage: %s file [file ...]\n", prog);
		return 1;
	}

	for(i=1; i<argc; i++) {

		if(argv[i][0] == '-')
			continue;

		puts("-------------------------------------------------------------------");
		puts(argv[i]);
		if((buff = readentire(argv[i], &size, 1500, prog)) != NULL) {
			print(buff, size);
			free(buff);
		} 
	}

	return 0;
}

/*==============================================================================
| print
|=============================================================================*/

void
print(u8 *buff, size_t size)
{
	machdr	*mac;
	size_t	offset = 14;

	mac = (machdr *) buff;

	puts("ethernet:");
	fputs("  type: ", stdout);

	if(mac->type == MACTYPE_IPDG)
		puts("IP datagram");
	else if(mac->type == MACTYPE_ARP)
		puts("ARP request/reply");
	else if(mac->type == MACTYPE_RARP)
		puts("RARP request/reply");
	else 
		printf("unknown [%04x]\n", mac->type);

	printf("  %02x:%02x:%02x:%02x:%02x:%02x => ",
	mac->src[0], mac->src[1], mac->src[2], 
	mac->src[3], mac->src[4], mac->src[5]);
	printf("%02x:%02x:%02x:%02x:%02x:%02x\n",
	mac->dst[0], mac->dst[1], mac->dst[2], 
	mac->dst[3], mac->dst[4], mac->dst[5]);

	if(mac->type == MACTYPE_IPDG) {
		offset += printip(buff + 14);
	}

	dumpdata(buff + offset, size - offset);
}

/*==============================================================================
| printip
|=============================================================================*/

size_t
printip(u8 *buff)
{
	iphdr		*ip = (iphdr *) buff;
	size_t	len = ip->ihl << 2; //IPIHL(ip->ihl_version) << 2;
	size_t	doff = len;

	puts("IP:");
	printf("  version:           %d\n", ip->version); //IPVER(ip->ihl_version));
	printf("  header length:     %02x\n", (int) len); 
	fputs ("  type of service:   ", stdout);

	switch(ip->tos) {
		case 0x00: puts("none"); break;
		case 0x10: puts("minimize delay"); break;
		case 0x08: puts("maximize thruput"); break;
		case 0x04: puts("maximize reliability"); break;
		case 0x02: puts("minimize monitary cost"); break;
		default:
	printf("unknown [%02x]\n", ip->tos);
	break;
	}

	printf("  total length       %04x\n", ntohs(ip->tot_len));
	printf("  frag id            %04x\n", ntohs(ip->id));
	printf("  frag offset        %04x\n", ntohs(ip->frag_off));
	printf("  time to live       %02x\n", ip->ttl);
	fputs ("  protocol           ", stdout);

	switch(ip->protocol) {
		case PROTOCOL_TCP: 
			puts("tcp"); 
			printtcp(buff, len);
			doff += sizeof(tcphdr);
			printaddrport(buff, len);
			break;

		case PROTOCOL_UDP:
			puts("udp");
			printaddrport(buff, len);
			doff += sizeof(udphdr);
			break;

		case PROTOCOL_ICMP:
			puts("icmp");
			printaddr(ip);
			printicmp(buff, len);
			doff += sizeof(icmphdr);
			break;

		default:
			printf("unknown [%02x]\n", ip->protocol);
			//printaddr(buff);
			break;

	}  

	return doff;
}

/*==============================================================================
| printaddrport
|=============================================================================*/

void
printaddrport(u8 *buff, size_t len)
{
	iphdr	*ip = (iphdr *) buff;
	udphdr *udp = (udphdr *) &buff[len];

	printf("  %s:%d => %s:%d\n",
	ip2string(ip->saddr), ntohs(udp->source),
	ip2string(ip->daddr), ntohs(udp->dest));
}

/*==============================================================================
| printtcp
|=============================================================================*/

void
printtcp(u8 *buff, size_t len)
{
	tcphdr *tcp = (tcphdr *) &buff[len];

	printf("  sequence number    %08x\n", (u32) ntohl(tcp->seq));
	printf("  ack number         %08x\n", (u32) ntohl(tcp->ack_seq));
	printf("  doff               %x\n", tcp->doff << 2);
	fputs ("  flags              ", stdout);
	if(tcp->fin) 
		fputs("fin ", stdout);
	if(tcp->syn) 
		fputs("syn ", stdout);
	if(tcp->rst) 
		fputs("rst ", stdout);
	if(tcp->psh) 
		fputs("psh ", stdout);
	if(tcp->ack) 
		fputs("ack ", stdout);
	if(tcp->urg) 
		fputs("urg", stdout);
	putchar('\n');
	printf("  window size        %04x\n", ntohs(tcp->window));
	printf("  urg pointer        %04x\n", ntohs(tcp->urg_ptr));
}

/*==============================================================================
| printaddr
|=============================================================================*/

void
printaddr(iphdr *ip)
{
	//iphdr	*ip = (iphdr *) buff;

	printf("  %s => %s\n", ip2string(ip->saddr), ip2string(ip->daddr));
}

/*==============================================================================
| printicmp
|=============================================================================*/

void
printicmp(u8 *buff, size_t len)
{
	icmphdr	*icmp = (icmphdr *) &buff[len];
	int		i,n=0;

	fputs ("  type:              ", stdout);
	for(i=0; i<MAXICMPTYPE; i++) {
		if(icmptype[i].index == icmp->type) {
			puts(icmptype[i].name);
			n = 1;
			break;
		}
	}
	if(n == 0) {
		printf("unknown [%02x]\n", icmp->type);
	}

	printf("  code:              %02d ", icmp->code);
	for(i=0; i<MAXICMPCODE; i++) {
		if(icmpcode[i].index == icmp->code) {
			fputs(icmpcode[i].name, stdout);
			putchar(' ');
		}
	}

	printf("\n  id:                %04x\n", ntohs(icmp->un.echo.id));
	printf("  sequence:          %04x\n", ntohs(icmp->un.echo.sequence));
	printf("  gateway:           %08x\n", (u32) ntohl(icmp->un.gateway));
}  

